Hotfix Makefile / error language / etc

Kopeerde opened the pull request:
- Changed JS files for Typescript ones.
- Updated .gitignore to not include node_modules.
- Makefile should properly work now.
- All 3 containers are written, now should link them and WIP on the backend.

Commit:
Hotfix Makefile / error language / etc

- Changed JS files for Typescript ones.
- Updated .gitignore to not include node_modules.
- Makefile should properly work now.
- All 3 containers are written, now should link them and WIP on the backend.

Bee-Have commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
-**/.env.private
-.notes
\ No newline at end of file
+**/.env
```
.env should not be in gitignore.
Not all env variables are sensitive informations (like path or ports and such)
There should be two env files : ".env" (no sensitive informations) ".env.private" (only the sensitive informations)

Maxou974 replied:
To me the best way is to have a .env_schema with all keys and non sensitive values and gitignoresthe .env, then we just have to copy the .env_schema in .env and add sensitive data

Kopeerde replied:
I recommend you reading the project subject :
> For obvious security reasons, any credentials, API keys, env
variables etc... must be saved locally in a .env file and ignored by
git.

therefore sensitive informations or not, it HAS to NOT be pushed on distant repo.

Bee-Have replied:
technically speaking, the subject says two things, first, ".env" should indeed be in the gitignore. However it also says that only "Publicly stored credentials" shouldn't be found in the project.
So i propose a workaround to make our life easier : ".env.something here to make it different" (this one will have our paths to make it easier) and another ".env" (this one has all the sensitive informations)
I also do not mind if we  have only one .env file which we do not push

Bee-Have commented on Makefile:
```diff
@@ -1,15 +1,33 @@
+enter_nginx:
+	docker exec -it nginx bash
 
-.PHONY: all logs fclean clean re
```
PHONY rule should be kept in the makefile. 
it is an important rule to have

Kopeerde replied:
i forgot to add them since i was in a hurry, new push fixed that issue.

Bee-Have replied:
No problem, that's why there are reviews, so we can all make sure to share the best work and learn.

Bee-Have commented on back_end/Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM node:latest
```
should not be "latest" but actual version.
If a new version suddenly gets out before the correction day we will be fucked.
it should be the latest version, but written in hard numbers

Kopeerde replied:
> You are free to use any library you want to in this context. However, you must use
the latest stable version of every library or framework used in your project

I know that some update might break things but the subjects requires to have latest versions, they say "library or framework" but better be safe than sorry.

Bee-Have replied:
Yes you are right, but wouldn't it be easier to change it in our own terms ?
A written version doesn't mean its not the latest no ?
We can keep this one if you think its wiser.

Bee-Have commented on docker-compose.yml:
```diff
@@ -0,0 +1,59 @@
+version: '3.8'
+
+volumes:
```
we should not have volumes. volumes create extra directories that are a bridge between the machine and the virtual container (binds mounts with extra steps).
Things we change will be updated in the volumes rather than our project directories.
Instead of volume it should be bind mounts. linking a physical directory in our project to a directory in the virtual container.

Kopeerde replied:
volumes are required for data persistence, plus you will not ship your actual pc, hence the bind between the volume and the directory, plus :
> volumes create extra directories that are a bridge between the machine and the virtual container
....... that's exactly what is needed, unless you want to rebuild your image every single time and have to wait 30 mins for the server to boot up instead of hot reaload (and no, you can ask echo, it took him arround 40 mins to build/start yesterday at building's pc).

Bee-Have replied:
Ok, if you think its better. I just thought bind mounts were interesting also. Let's use volumes.

Bee-Have commented on docker-compose.yml:
```diff
@@ -0,0 +1,59 @@
+    networks:
+      - transc_netw
+    ports:
+      - 5432:5432
```
ports should be added to the public ".env" file as to avoid errors.
it is incredibly easy to miss click and write "5431" instead of "5432".
Especially since this information will be present on multiple files.

Kopeerde replied:
the issue here is a non issue, anyway, changed it because it makes easier to change at multiple parts (and NOT because you can missclick, that is a non issue).

Bee-Have commented on docker-compose.yml:
```diff
@@ -0,0 +1,59 @@
+services:
+  postgres:
+    container_name: postgres
+    volumes:
```
bind mount here instead of volume

Kopeerde replied:
like i previous said, i recommend you to read the subject :

> When your computers in clusters run under Linux, you will use
Docker in rootless mode for security reasons. This comes with 2
sideways [...] you can’t use so called “bind-mount volumes”
between the host and the container if non-root UIDs are used in the
container

Therefore you tell me to use bind-mounts while being unable to use bind mounts.

Bee-Have replied:
I indeed didn't remember that part, my bad, mistakes happen, we're students after all.

Bee-Have commented on docker-compose.yml:
```diff
@@ -0,0 +1,59 @@
+    depends_on:
+      - postgres
+    container_name: front_end
+    volumes:
```
again, not volume but bind mount.
Whenever the container will be changed, we want the changes to be kept and it is better to use bind mounts

Kopeerde replied:
same as previously.

Bee-Have commented on docker-compose.yml:
```diff
@@ -0,0 +1,59 @@
+      - CHOKIDAR_USEPOLLING=true
+      - WATCHPACK_POLLING=true
+    ports:
+      - 3000:3000
```
ports should be in .env file

Kopeerde replied:
fixed in next push.

Bee-Have commented on docker-compose.yml:
```diff
@@ -0,0 +1,59 @@
+      - postgres
+    container_name: back_end
+    volumes:
+      - html_vol:/usr/src/app/frontend
```
volumes should be bind mounts

Kopeerde replied:
once again already talked about volumes.

Bee-Have commented on front_end/Dockerfile:
```diff
@@ -1,21 +1,24 @@
-FROM node:20.10.0
+FROM node:latest
```
should not be "latest" but actual version.
If a new version suddenly gets out before the correction day we will be fucked.
it should be the latest version, but written in hard numbers

Kopeerde replied:
already talked about it

Bee-Have commented on front_end/Dockerfile:
```diff
@@ -1,21 +1,24 @@
 
-RUN mkdir -p /usr/src/app/frontend
+# expose port
+EXPOSE 3000
```
port should be in env file to avoid mistakes

Kopeerde replied:
it should be use in .env in case it's NOT in dockerfile since the docker engine will create an image using default environment, the .env file is only used while running the container, meaning that writing :

EXPOSE ${PORT_POSTGRESS} is the same as EXPOSE ""

Unless you export the variable, which defeats the purpose of the .env file.

Bee-Have replied:
You can pass variables in Dockerfiles using the docker-compose file. It's quite handy.

Bee-Have commented on Makefile:
```diff
@@ -1,15 +1,33 @@
 
-clean:
-	cd srcs && docker-compose down --remove-orphans
+$(NAME): build
```
This rule call depends on the order of the makefile. It is prone to error upon editing.
This rule shouldn't have to be called directly or by default.
There should always be a "all" rule. 
Therefore you should have a "all" rule that calls $(NAME)

Kopeerde replied:
if two rules does exactly the same (here you say that all should call 
/$(NAME) and nothing else, then all == $(NAME) therefore, that rant you just made is useless.
But i'll add it anyway.

Bee-Have replied:
I did not intended it as a rent, i apologize,
having an extra rule in c/c++ usually is used to check if your executable is build or not.
Here since there is none, what might be easier, is to create a file upon building and checking if it exists.
Additionally this file could  be a log file to have an extra use.

Bee-Have commented on Makefile:
```diff
@@ -1,15 +1,33 @@
+	docker-compose down
+
+re: fclean
+	make $(NAME)
```
This should not exist. This calls another makefile.
THIS IS A RECURSIVE MAKEFILE. REALLY BAD.

Kopeerde replied:
So bad that it works and even the official documentation says that you can do it, anyway, gonna change it into a dependency.

Bee-Have commented on Makefile:
```diff
@@ -1,15 +1,33 @@
+re: fclean
+	make $(NAME)
+
+violence:
```
shouldn't it be called "prune" since it just calls prune ?

Kopeerde replied:
i didn't had time to change, i had to quickly push something for people to finally be able to work, anyway, fixed it in new push

Bee-Have replied:
I understand, no problem

Bee-Have commented on Makefile:
```diff
@@ -1,15 +1,33 @@
+violence:
+	docker system prune -af --all
+
+enter_postgres:
```
should be called "enter_database"

Kopeerde replied:
no, "database" is too vague, plus this container only runs postgress and in work environment (so not this project), you might setup a SQL database (like postgres) and a cache database (like redis) , therefore calling the rule "enter_database" would enter in which container?

Bee-Have replied:
I did not think of that, its a very good point.

Bee-Have commented on Makefile:
```diff
@@ -1,15 +1,33 @@
+enter_front_end:
+	docker exec -it front_end bash
+
+enter_nginx:
```
should be called "enter_back_end"

Kopeerde replied:
once again, good issue, wrong reason, had to push it for people to work, but if you looked a little bit more ,you'll see that nginx is not there (hence removed) and the command in the makefile therefore does not work.
Anyway, fixed in nex push

Bee-Have commented on postgres/Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM postgres:latest
```
This should not be latest but hardcoded version.
If the version changes we're fucked. So we should always change the version ourselves.
It is more professional for a project.

Kopeerde replied:
once again, same issue that previously talked.

Bee-Have commented on postgres/Dockerfile:
```diff
@@ -0,0 +1,7 @@
+
+RUN apt-get update
+
+EXPOSE 5432
```
port should bein ".env" file

Kopeerde replied:
already talked about it.

Bee-Have reviewed (changes requested):
This pull request feels more like an overwriting rather than an actual update.
As a result there were some architectural changes that render my work obsolete. It is not a problem, i do not mind deleting my work. 
HOWEVER i do not want useless directories lying around. So it will require additional cleanup.

Commit:
Fixing some issues from Makefile and Docker-compose

- Fixed issue with enter_nginx by changing into enter_back_end
- Changed port usage in Docker-compose.yml to use the .env file

Kopeerde commented:
Because the main goal was a hotfix (hense the name), since code did not worked at all, it had to be overwrite.
HOWEVER, i will not tolerate any bullshit PR, i made some mistakes, but 90% of what you said is wrong, plus the "useless" directories are useless because people will work on backend docker and nest js?? since when making the base of their work is useless??

Bee-Have commented:
Hi, I understand it had to be overwritten, it was not said in your PR that this was an overwriting so i thought i would include it.
I understand that i can be wrong about some things, we are students after all and mistakes are part of learning.
The useless directories i was talking about, were mine, my old work, not what you implemented, i apologize if you misunderstood my phrasing
